sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -1282,7 +1282,16 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     val functionName = funcDefinition.identifier.funcName.toLowerCase(Locale.ROOT)
     requireFunctionExists(db, functionName)
     val functionIdentifier = funcDefinition.identifier.copy(funcName = functionName)
-    client.alterFunction(db, funcDefinition.copy(identifier = functionIdentifier))
+    val newDefinition = funcDefinition.copy(identifier = functionIdentifier)
+    isResourceChanged(db, newDefinition) match {
+      // as of now Hive only alter name, owner, class name, type but not resource URI
+      // So in order to replace function with resource spark needs to delete and create the function
+      case true => client.dropFunction(db, functionName)
+        client.createFunction(db, newDefinition)
+      // replace the function in the metastore if there is no change in the resource
+      case _ =>
+        client.alterFunction(db, newDefinition)
```

Comment:
From Hive [DDL](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Create/Drop/ReloadFunction), looks like it does not have "replace" in create function:

```
CREATE FUNCTION [db_name.]function_name AS class_name
  [USING JAR|FILE|ARCHIVE 'file_uri' [, JAR|FILE|ARCHIVE 'file_uri'] ];
```

Maybe it is why this API does not do resource replacement?

Reply:
yes, as of now Hive only alter name, owner, class name, type but not resource URI. There is no replace DDL in Hive

Reply:
And also initially in the [jira](https://issues.apache.org/jira/browse/SPARK-28710), I proposed 2 solutions for this bug

solution 1: throw Unsupported Error for permanent function
solution 2: instead of alter function, do drop and create